### PR TITLE
overload: don't print empty  messages

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: Log messages during overload can be a bit spammy, and perhaps even contribute to the overload. Many of these messages I've seen in production have been empty (printing that nothing happened). This PR skips those.

You can consider this an RFC; if you think this is is not a good idea or have a better idea about how to handle this, please let me know; we can revert.
Additional Description:
Risk Level: low
Testing: none, just CI
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/common/buffer/watermark_buffer.cc [hunk 1]
```diff
@@ -210,18 +210,27 @@ uint64_t WatermarkBufferFactory::resetAccountsGivenPressure(float pressure) {
       std::floor(pressure * BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_) + 1, 8);
 
   uint32_t last_bucket_to_clear = BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_ - buckets_to_clear;
-  ENVOY_LOG_MISC(warn, "resetting streams in buckets >= {}", last_bucket_to_clear);
+  bool need_message = false;
```
Comment (KBaichoo) [inline source/common/buffer/watermark_buffer.cc:213]: this value is always false ?

File: source/common/buffer/watermark_buffer.cc [hunk 2]
```diff
@@ -210,18 +210,27 @@ uint64_t WatermarkBufferFactory::resetAccountsGivenPressure(float pressure) {
...
-    ENVOY_LOG_MISC(warn, "resetting {} streams in bucket {}.",
-                   size_class_account_sets_[bucket_to_clear].size(), bucket_to_clear);
+    absl::flat_hash_set<BufferMemoryAccountSharedPtr>& bucket =
+        size_class_account_sets_[bucket_to_clear];
+    if (bucket.empty()) {
+      // Don't print a message if there is nothing to clear.
+      continue;
+    }
+    if (need_message) {
+      need_message = false;
```
Comment (KBaichoo) [inline source/common/buffer/watermark_buffer.cc:227]: I think what you're trying to do here is print this message one per call if we have a bucket with some entity in it.

Consider inverting the logic e.g. 
```
bool printed_message = false; 
... // skip empty buckets
if(!printed_message) {
 printed_message = true;
 warn(...)
}
```

You could also consider having a log at the end like `ENVOY_LOG_EVERY_NTH` for if the buckets were empty every say 200 calls.

**Review by KBaichoo — COMMENTED** [review]
Seems reasonable to avoid spamming if nothing to reset.

/wait

File: source/common/buffer/watermark_buffer.cc (hunk 1 above)
Comment (jmarantz) [reply source/common/buffer/watermark_buffer.cc]: urr, done.

File: source/common/buffer/watermark_buffer.cc (hunk 2 above)
Comment (jmarantz) [reply source/common/buffer/watermark_buffer.cc:228]: LOG_EVERY_NTH is a good idea; done.

I don't see why the inverted logic is better.

**Comment by jmarantz** [comment]
Hi Kevin -- sorry for the delay; spammy messages came up for us again (100s of messages within 1/100 of a sec).

You up for continuing this review?

**Comment by jmarantz** [comment]
Actually hold off a bit I want to add tests. Unfortunately the code I'm editing does not fire in a unit test but it does fire in test/integration/buffer_accounting_integration_test, method `ResetsExpensiveStreamsWhenUpstreamBuffersTakeTooMuchSpaceAndOverloaded`.

**Comment by KBaichoo** [comment]
will review when this is ready

**Comment by jmarantz** [comment]
I'm going to re-open though I couldn't make a robust test as the reset behavior is not unit tested at all, and the integration test behavior does not appear to be deterministic.

**Review by KBaichoo — APPROVED** [review]
Thanks, I think this will be more condensed but might still print empty in some cases.

**Review by KBaichoo — CHANGES_REQUESTED** [review]
Can you add a test?

**Comment by jmarantz** [comment]
Oh thanks -- I had missed that. I had just verified that the logging line did not get hit from watermark_buffer_test.cc and didn't notice there was another test file.

**Comment by jmarantz** [comment]
good catch on the remaining empty messages; added a test for that too.

**Review by KBaichoo — APPROVED** [review]

**Pull request merged**